fl_training.py

```diff
@@ -38,88 +36,110 @@ def preprocess_node_list(node_list):
     """Return the score on test data of a model trained on a single node"""
     
+    print('\n## Training and evaluating model on one single node.')
     # Initialize model
     model = utils.generate_new_cnn_model()
     # print(model.summary())
 
     # Train model
-    print('\n### Training model on one single node: node ' + node.node_id)
+    print('\n### Training model on one single node: ' + str(node))
     history = model.fit(node.x_train, node.y_train,
               batch_size=constants.BATCH_SIZE,
               epochs=epoch_count,
               verbose=0,
               validation_data=(node.x_val, node.y_val))
     
     # Evaluate trained model
-    print('\n### Evaluating model on test data:')
+    print('\n### Evaluating model on test data of the node:')
     model_evaluation = model.evaluate(node.x_test, node.y_test,
                            batch_size=constants.BATCH_SIZE,
                            verbose=0)
-    print('\nModel metrics names: ', model.metrics_names)
-    print('Model metrics values: ', ['%.3f' % elem for elem in model_evaluation])
+    print('   Model metrics names: ', model.metrics_names)
+    print('   Model metrics values: ', ['%.3f' % elem for elem in model_evaluation])
     
     model_eval_score = model_evaluation[1] # 0 is for the loss
 
     # Return model score on test data
+    print('\nTraining and evaluation on one single node: done.')
     return model_eval_score
 
-# TODO no methods overloadin
+
+#%% TODO no methods overloading
 def compute_test_score_with_scenario(scenario, is_save_fig=False):
-    return compute_test_score(scenario.node_list, 
-                              scenario.epoch_count, 
+    return compute_test_score(scenario.node_list,
+                              scenario.epoch_count,
+                              scenario.x_esval,
+                              scenario.y_esval,
+                              scenario.x_test,
+                              scenario.y_test,
                               scenario.is_early_stopping,
                               is_save_fig,
                               save_folder=scenario.save_folder)
         
         
-# Distributed learning training      
-def compute_test_score(node_list, epoch_count, is_early_stopping=True, is_save_fig=False, save_folder=''):
+#%% Distributed learning training      
+def compute_test_score(node_list, epoch_count, x_esval, y_esval, x_test, y_test, is_early_stopping=True, is_save_fig=False, save_folder=''):
     """Return the score on test data of a final aggregated model trained in a federated way on each node"""
 
     nodes_count = len(node_list)
-        
+    
+    # If only one node, fall back to dedicated single node function    
     if nodes_count == 1:
         return compute_test_score_for_single_node(node_list[0], epoch_count)
     
+    # Else, follow a federated learning procedure
     else:
+        print('\n## Training and evaluating model on multiple nodes: ' + str(node_list))
```

Comment:
This is the output on my machine:
> ## Training and evaluating model on multiple nodes: [<node.Node object at 0x000001F347E2B748>, <node.Node object at 0x000001F347DCEF98>]
Because Node class does not have` __str__` or `__repr__` methods implemented --> https://stackoverflow.com/questions/4932438/how-to-create-a-custom-string-representation-for-a-class-object

Reply:
Yes I know. I couldn't find an easy way to give a sort of a "name" or an "id" that sticks to a given node, and I figured that printing this memory address was not that bad.

Reply:
I agree. Then we should not print this line since it does not provide any useful information. Or juste print the length of the node list?